crates/katana/core/src/backend/gas_oracle.rs

```diff
@@ -1,25 +1,282 @@
+use std::backtrace::Backtrace;
+use std::collections::VecDeque;
+use std::fmt::Debug;
+use std::future::IntoFuture;
+use std::pin::Pin;
+
+use alloy_provider::{Provider, ProviderBuilder};
+use alloy_rpc_types_eth::BlockNumberOrTag;
+use alloy_transport::Transport;
+use anyhow::{Context, Ok};
+use futures::Future;
 use katana_primitives::block::GasPrices;
+use tokio::time::Duration;
+use url::Url;
+
+const BUFFER_SIZE: usize = 60;
+const INTERVAL: Duration = Duration::from_secs(60);
+const ONE_GWEI: u128 = 1_000_000_000;
 
 // TODO: implement a proper gas oracle function - sample the l1 gas and data gas prices
 // currently this just return the hardcoded value set from the cli or if not set, the default value.
 #[derive(Debug)]
-pub struct L1GasOracle {
+pub enum L1GasOracle {
+    Fixed(FixedL1GasOracle),
+    Sampled(SampledL1GasOracle),
+}
+
+#[derive(Debug)]
+pub struct FixedL1GasOracle {
     gas_prices: GasPrices,
     data_gas_prices: GasPrices,
 }
 
+#[derive(Debug, Default, Clone)]
+pub struct SampledL1GasOracle {
+    gas_prices: GasPrices,
+    data_gas_prices: GasPrices,
+}
```

Comment:
We need to use the same instance of the gas prices for both `SampledL1GasOracle` and the  `GasOracleWorker`. Right now, the values that is collected by the worker is not being shared with the one in the `Backend`. 

https://github.com/dojoengine/dojo/blob/7d6b7d91b710ba0c9b8ec66c132be60eae41feaf/crates/katana/core/src/backend/mod.rs#L44

thus, the gas prices returned here would always be zero:

https://github.com/dojoengine/dojo/blob/7d6b7d91b710ba0c9b8ec66c132be60eae41feaf/crates/katana/core/src/backend/mod.rs#L112-L122

as `GasOracleWorker` is only updating to its own instance of `SampledL1GasOracle`, which the `Backend` has no shared access to.

https://github.com/dojoengine/dojo/blob/7d6b7d91b710ba0c9b8ec66c132be60eae41feaf/crates/katana/core/src/backend/gas_oracle.rs#L142-L151

We can wrap a shared inner type like this:

```rust
#[derive(Debug, Default, Clone)]
pub struct SampledL1GasOracle {
    prices: Arc<Mutex<SampledPrices>>,
}

#[derive(Debug, Default)]
struct SampledPrices {
    gas_prices: GasPrices,
    data_gas_prices: GasPrices,
}

#[derive(Debug, Clone)]
pub struct GasOracleWorker {
    prices: Arc<Mutex<SampledPrices>>,
    l1_provider_url: Option<Url>,
}

impl SampledL1GasOracle {
    pub fn current_data_gas_prices(&self) -> GasPrices {
        self.prices.lock().data_gas_prices.clone()
    }

    pub fn current_gas_prices(&self) -> GasPrices {
        self.prices.lock().gas_prices.clone()
    }
}
```

and the worker will update the shared `SampledPrices`